### PR TITLE
Fix Orca Warning / Error Message

### PR DESCRIPTION
**Pull request by karthijrk**
Using elog::ERROR in GPDB translator might result in a memory leak, as it jump immediately to PostgresMain. Hence when an error happens we print the message as a warning and throw C++ exception to unwind the call stack. But seeing warning followed by the error message is confusing to the user.

E.g.

```
WARNING:  null value in column "col3" violates not-null constraint
-ERROR:  PQO unable to generate a plan, please see the above message for details. (CGPOptimizer.cpp:66)
```

In this PR, we introduce the specific exception code for those cases and called elog::ERROR in the last call stack to avoid a memory leak. With this changes, user now see the same error as shown below.

E.g. 

```
ERROR:  PQO unable to generate plan, null value in column "col3" violates not-null constraint. (CGPOptimizer.cpp:82)
```

@vraghavan78 @foyzur @hardikar @armenatzoglou Please take a look when you get chance.

### CODE REVIEW

File: src/backend/gpopt/CGPOptimizer.cpp [hunk 1]
```diff
@@ -57,23 +60,50 @@ CGPOptimizer::PplstmtOptimize
...
 	{
+		// copy the error message before context free.
+		CHAR* szErrorMsg = NULL;
+		if (NULL != octx.m_szErrorMsg)
+		{
+			szErrorMsg = gpdb::SzMemoryContextStrdup(MessageContext, octx.m_szErrorMsg);
+		}
+		// clean up context
+		octx.Free(octx.epinQuery, octx.epinPlStmt);
 		if (GPOS_MATCH_EX(ex, gpdxl::ExmaDXL, gpdxl::ExmiWarningAsError))
```
Comment (karthijrk) [inline src/backend/gpopt/CGPOptimizer.cpp]: @vraghavan78  Please let me know the error message in this `catch` blocks looks good from user perspective. Please refer the regression output for some examples in this PR.

**Comment by karthijrk** [comment]
Corresponding ORCA changes - https://github.com/greenplum-db/gporca/pull/113

**Comment by hlinnaka** [comment]
I would drop the "PQO unable to generate a plan" part from the error messages. Based on the examples in the regression suite, the errors have nothing to do with planning per se. It's not particularly interesting to the user that they errors occurred while planning. And I assume if you were not using ORCA, that's what you'd get anyway, so it would be good to be consistent.

Also, if I'm reading the diff correctly, we are losing the error code and any other information from the original error. E.g. errcode(ERRCODE_SYNTAX_ERROR). That determines the SQLSTATE of the error, which is visible to clients, which may take different actions depending on the kind of error. As well as the error context stack, etc. That's quite unfortunate.

Instead of constructing a new error, should use something like pg_re_throw() to re-throw the original one.

**Comment by karthijrk** [comment]
@hlinnaka  You mean we loose the errcode because we are rewriting the error message ?

**Comment by vraghavan78** [comment]
@karthijrk, I think keeping error messages between GPORCA and Planner similar is good. So I like @hlinnaka 's suggestion. Please do the needful.

**Comment by karthijrk** [comment]
Great. I will change it accordingly. In this case, I don't need to construct a new error message too. :)

**Comment by hlinnaka** [comment]
On 10/25/2016 10:25 PM, Karthikeyan Jambu Rajaraman wrote:
> @hlinnaka  You mean we loose the errcode because we are rewriting the error message ?

Right.

- Heikki

**Comment by karthijrk** [comment]
@vraghavan78 @foyzur @hlinnaka  FYI - I addressed all the PR comments. Please have a look.

File: src/backend/gpopt/utils/COptTasks.cpp [hunk 2]
```diff
@@ -1667,26 +1657,27 @@ PlannedStmt *
...
 	GPOS_CATCH_EX(ex)
 	{
-		octx.HandleError(pfUnexpectedFailure);
+		octx->HandleError(pfUnexpectedFailure);
 		GPOS_RETHROW(ex);
 	}
 	GPOS_CATCH_END;
-	octx.HandleError(pfUnexpectedFailure);
-	return octx.m_pplstmt;
+	octx->HandleError(pfUnexpectedFailure);
```
Comment (karthijrk) [inline src/backend/gpopt/utils/COptTasks.cpp]: I thought there may be a case where we might not throw an exception but have an error message in octx. The code before this change has the same behavior. So to be safer I preserved the same.

File: src/include/gpopt/utils/COptTasks.h [hunk 3]
```diff
@@ -51,66 +51,64 @@ using namespace gpos;
...
 
-class COptTasks
+// context of optimizer input and output objects
+struct SOptContext
 {
 
-	private:
+	// mark which pointer member should NOT be released
+	// when calling Free() function
+	enum EPin
```
Comment (foyzur) [inline src/include/gpopt/utils/COptTasks.h]: I know @vraghavan78 will cringe. But, didn't we decide that for new code we can keep non-hungarian? It's difficult to read even for code review :-(

Comment (foyzur) [inline src/include/gpopt/utils/COptTasks.h]: Oh I see it's old code but updated formatting.

Comment (karthijrk) [inline src/include/gpopt/utils/COptTasks.h]: Oh, I forget about this completely. That's good we don't need to use Hungarian any more. BTW, this is not a new code. This struct was nested in COptTasks as a private struct. I moved this out because this struct has to be used in CGPOptimizer.

File: src/test/regress/expected/role_optimizer.out [hunk 4]
```diff
@@ -1,107 +0,0 @@
---
```
Comment (foyzur) [inline src/test/regress/expected/role_optimizer.out]: Why is this file deleted?

File: src/test/regress/expected/privileges_optimizer.out [hunk 5]
```diff
@@ -1,733 +0,0 @@
---
```
Comment (foyzur) [inline src/test/regress/expected/privileges_optimizer.out]: Why is this file deleted?

File: src/test/regress/expected/namespace_optimizer.out [hunk 6]
```diff
@@ -1,142 +0,0 @@
---
```
Comment (foyzur) [inline src/test/regress/expected/namespace_optimizer.out]: Why is this file deleted?

**Comment by karthijrk** [comment]
Earlier there was a difference in how the error message is printed in orca. Now they are same we don't need to have separate out file for optimizer. Hence those files are deleted.

**Comment by karthijrk** [comment]
Brief of latest change in this PR 

In ORCA, due to the way exception handled previously we do warning first
and then later print error referring that message. In this commit, we
enhanced the exception handling so we just print a single error message.
Also, we remove 'PQO unable to generate a plan' or 'Aborting PQO plan
generation' message and make the error message as close as the planner.

**Comment by karthijrk** [comment]
@oarap @d Please take a look when you guys get chance.

File: src/backend/gpopt/CGPOptimizer.cpp [hunk 7]
```diff
@@ -57,23 +60,44 @@ CGPOptimizer::PplstmtOptimize
...
```
Comment (oarap) [inline src/backend/gpopt/CGPOptimizer.cpp]: `ExmiWarningAsError` should be removed after this refactoring, since we address all the errors here.

File: src/backend/gpopt/utils/COptTasks.cpp [hunk 8]
```diff
@@ -167,30 +167,21 @@ COptTasks::SOptContext::SOptContext()
...
-		elog(WARNING, "%s", m_szErrorMsg);
-	}
 	*pfUnexpectedFailure = m_fUnexpectedFailure;
-
-	// clean up context
-	Free(epinQuery, epinPlStmt);
-	if (bhasError)
+	if (NULL != m_szErrorMsg)
 	{
 		GPOS_RAISE(gpdxl::ExmaDXL, gpdxl::ExmiWarningAsError);
```
Comment (oarap) [inline src/backend/gpopt/utils/COptTasks.cpp]: How about when we raise the exception, copy the error message to the exception as exception message, instead of using the `SOptContext` to pass it back? e.g.

`GPOS_RAISE(gpdxl::ExmaDXL, gpdxl::ExmiOptimizerError, m_szErrorMsg)`

`ExmiOptimizerError` is used instead of `ExmiWarningAsError`.

File: src/backend/gpopt/utils/COptTasks.cpp [hunk 9]
```diff
@@ -376,8 +367,7 @@ COptTasks::SzAllocate
 	}
 	GPOS_CATCH_EX(ex)
 	{
-		elog(WARNING, "no available memory to allocate string buffer");
-		GPOS_RAISE(gpdxl::ExmaDXL, gpdxl::ExmiWarningAsError);
```
Comment (oarap) [inline src/backend/gpopt/utils/COptTasks.cpp]: Maybe just 

```
GPOS_RAISE(gpdxl::ExmaDXL, gpdxl::ExmiOptimizerError, GPOS_WSZ_LIT("no available memory to allocate string buffer"));
```

File: src/backend/gpopt/utils/COptTasks.cpp [hunk 10]
```diff
@@ -1898,8 +1889,7 @@ COptTasks::UlCmpt
 		}
 	}
 
-	elog(WARNING, "Invalid comparison type code. Valid values are Eq, NEq, LT, LEq, GT, GEq");
-	GPOS_RAISE(gpdxl::ExmaDXL, gpdxl::ExmiWarningAsError);
```
Comment (oarap) [inline src/backend/gpopt/utils/COptTasks.cpp]: Maybe just:

```
GPOS_RAISE(gpdxl::ExmaDXL, gpdxl::ExmiOptimizerError, GPOS_WSZ_LIT("Invalid comparison type code. Valid values are Eq, NEq, LT, LEq, GT, GEq"));
```

File: src/backend/gpopt/CGPOptimizer.cpp (hunk 7 above)
Comment (oarap) [inline src/backend/gpopt/CGPOptimizer.cpp]: Maybe the whole block can be like:

```
if (GPOS_MATCH_EX(ex, gpdxl::ExmaGPDB, gpdxl::ExmiGPDBError))
{
    PG_RE_THROW();
}
else if (GPOS_MATCH_EX(ex, gpdxl::ExmaGPDB, gpdxl::ExmiOptimizerError))
{
    elog(ERROR, "PQO unable to generate plan due to: %s, call stack %s", "/* some how we get the error message here from the exception */", "/* get the original call stack */");
}
```

File: src/backend/gpopt/utils/COptTasks.cpp (hunk 9 above)
Comment (karthijrk) [inline src/backend/gpopt/utils/COptTasks.cpp]: I am afraid whether GPOS_RAISE works ? It keeps the error message in the Task's error context and we don't have access to context when we want to do elog::ERROR. I thought for the same reason translator had SOptContext to get the error message in that object.

File: src/backend/gpopt/utils/COptTasks.cpp (hunk 10 above)
Comment (karthijrk) [inline src/backend/gpopt/utils/COptTasks.cpp]: For same reason as mentioned earlier, I am not sure if we can do this.

File: src/backend/gpopt/utils/COptTasks.cpp [hunk 11]
```diff
@@ -147,13 +147,13 @@ const ULONG rgulExpectedDXLErrors[] =
 
 //---------------------------------------------------------------------------
 //	@function:
-//		COptTasks::SOptContext::SOptContext
+//		SOptContext::SOptContext
```
Comment (karthijrk) [inline src/backend/gpopt/utils/COptTasks.cpp]: What advantage we gain by moving it as class ? It was a struct before inside COptTasks.

File: src/backend/gpopt/CGPOptimizer.cpp [hunk 12]
```diff
@@ -57,23 +60,44 @@ CGPOptimizer::PplstmtOptimize
...
+			Assert(NULL != szErrorMsg);
+			elog(ERROR, "%s", szErrorMsg);
+		}
+		else if (GPOS_MATCH_EX(ex, gpdxl::ExmaGPDB, gpdxl::ExmiGPDBError))
+		{
+			PG_RE_THROW();
+		}
+		else if (GPOS_MATCH_EX(ex, gpdxl::ExmaGPDB, gpdxl::ExmiNoAvailableMemory))
+		{
+			elog(ERROR, "No available memory to allocate string buffer.");
```
Comment (karthijrk) [inline src/backend/gpopt/CGPOptimizer.cpp]: I agree that we lose the line number. 
I wonder how to get the line number into the error message. FYI, we can't do elog::ERROR in the place where the allocation failed. May we can talk in offline and see if we there is any other way to do it.

File: src/backend/gpopt/CGPOptimizer.cpp (hunk 7 above)
Comment (karthijrk) [inline src/backend/gpopt/CGPOptimizer.cpp]: We want to keep the error message same as in planner. That's the whole purpose of this PR. Also we don't have call stack by now because we already unwinded the call stack. We can talk in offline more.

File: src/include/gpopt/utils/COptTasks.h [hunk 13]
```diff
@@ -51,66 +51,64 @@ using namespace gpos;
 using namespace gpdxl;
 using namespace gpopt;
 
-class COptTasks
+// context of optimizer input and output objects
+struct SOptContext
```
Comment (karthijrk) [inline src/include/gpopt/utils/COptTasks.h]: Let's discuss this in offline.

File: src/backend/gpopt/CGPOptimizer.cpp (hunk 7 above)
Comment (karthijrk) [inline src/backend/gpopt/CGPOptimizer.cpp]: I think we still need this enum. Maybe can be renamed to ExmiOptimizerError.

File: src/backend/gpopt/utils/COptTasks.cpp (hunk 9 above)
Comment (xinzweb) [inline src/backend/gpopt/utils/COptTasks.cpp]: Yeah, that's the context we (@oarap and I) are still learning on. E.g., seems like there is no easy way pass the error message along the exceptions, but rely on the `SOptContext`. 

If we just push one more step, is that possible to refactor the `CException` to add this error message context as the `SOptContext` captured?

File: src/backend/gpopt/utils/COptTasks.cpp (hunk 10 above)
Comment (karthijrk) [inline src/backend/gpopt/utils/COptTasks.cpp]: Same as above

File: src/backend/gpopt/utils/COptTasks.cpp (hunk 8 above)
Comment (karthijrk) [inline src/backend/gpopt/utils/COptTasks.cpp]: We can't do it because the error message will be lost because current CTask is null.

File: src/include/gpopt/utils/COptTasks.h (hunk 13 above)
Comment (karthijrk) [inline src/include/gpopt/utils/COptTasks.h]: Can we have this as a separate chores. This requires lots of changes because in many places we access its field directly. Moving to class will make it private and need public member function to expose the access.

**Comment by karthijrk** [comment]
@oarap @xinzweb I address all your PR comments. Also proposing to create a chore for changing SOptContext to a class in the separate file. This requires quite a few changes and new public member functions. 

Please have a lot look when you get chance.

**Comment by karthijrk** [comment]
Thanks a lot guys for reviewing it.
